tests/FSharp.Editing.Tests/GoToDefinitionTests.fs

```diff
@@ -124,56 +128,98 @@ let generateDefinitionFromPos caretPos src =
-
-        // Implicit 'new' statement: symbol is considered as a function
-        """open System
-let x = Tuple<int, int>(1, 2)""", (Pos.fromZ 1 8)
-    ]
-    |> List.map (fun (src, pos) -> generateDefinitionFromPos pos src)
-    |> List.iter (fun src ->
-        assertSrcAreEqual src
-            """namespace System
-
-/// Represents a 2-tuple, or pair. 
-type Tuple<'T1, 'T2> =
-    /// Initializes a new instance of the  class.
-    new : item1:'T1 * item2:'T2 -> Tuple<'T1, 'T2>
-    /// Returns a value that indicates whether the current  object is equal to a specified object.
-    member Equals : obj:obj -> bool
-    /// Returns the hash code for the current  object.
-    member GetHashCode : unit -> int
-    /// Gets the value of the current  object's first component.
-    member Item1 : 'T1
-    /// Gets the value of the current  object's second component.
-    member Item2 : 'T2
-    /// Returns a string that represents the value of this  instance.
-    member ToString : unit -> string
-""")
-
-[<Test>]
-let ``adds necessary parenthesis to function parameters`` () =
-    """open System
-let _ = Async.AwaitTask"""
-    |> generateDefinitionFromPos (Pos.fromZ 1 8)
-    |> assertSrcAreEqualForFirstLines 10 """namespace Microsoft.FSharp.Control
-
-open System
+[<NoComparison>]
+type GenericTest =
+| Skip of reason: string
+| AssertEqual of input: string * position: pos * expected: string
+| AssertEqualFirstLines of input: string * position: pos * expected: string * firstLines: int
+
+[<NoComparison>]
+type GenericTestResult =
+| Passed
+| Inconclusive
+| Failed of exn
+| IncorrectFolderSetup of message: string
+| IncorrectResult of message: string
+with
+  static member HasFailed = function | Passed | Inconclusive -> false | _ -> true
+
+let getNLines n = 
+  String.getLines 
+  >> Array.truncate n 
+  >> String.concat Environment.NewLine
+
+[<Test>]
+let ``run generic tests`` () =
+  
+  let parseTest (directory: DirectoryInfo) =
+    let source   = directory.FullName </> "input.fs"      |> File.ReadAllText
+    let expected = directory.FullName </> "expected.fs"   |> File.ReadAllText
+    let settings = directory.FullName </> "settings.json" |> GenericTestSettings.Load
+    let caretPos = mkPos settings.Pos.Line settings.Pos.Col
+    let ignoreOnUnix = 
+      match settings.IgnoreOnUnix with
+      | Some true -> true
+      | _ -> false
+    if ignoreOnUnix && PlatformID.Win32NT <> Environment.OSVersion.Platform then
+      Skip "ignore on unix"
+    else
+    match settings.AssertType with
+    | "eq"           -> AssertEqual(source, caretPos, expected)
+    | "firstLinesEq" -> AssertEqualFirstLines(source, caretPos, expected, settings.FirstLineCount.Value)
+    | _ -> failwithf "assertType '%s' is invalid, expected 'eq' or 'firstLinesEq'" settings.AssertType
+
+  let executeTest directory =
+    let test =
+      try
+        Choice1Of2 <| parseTest directory
+      with
+      | e -> Choice2Of2 (string e)
```

Comment:
As we started to use Either, let's introduce `Ok` and `Fail` aliases and `Ok` and `Fail` APs for completeness.